src/github.com/getlantern/flashlight/localdiscovery/localdiscovery.go

```diff
@@ -0,0 +1,90 @@
+// Service for discovering Lantern instances in the local network
+package localdiscovery
+
+import (
+	"sync"
+	"time"
+
+	"github.com/getlantern/flashlight/ui"
+	"github.com/getlantern/golog"
+	"github.com/getlantern/multicast"
+)
+
+const (
+	messageType  = `LocalDiscovery`
+	updatePeriod = 10
+)
+
+var (
+	log           = golog.LoggerFor("flashlight.localdiscovery")
+	service       *ui.Service
+	mc            *multicast.Multicast
+	lastPeers     []multicast.PeerInfo
+	peersMutex    sync.Mutex
+)
+
+func Start(advertise bool, portToAdvertise string) {
+	if service != nil {
+		// Dev error: this service shouldn't be started unless stopped
+		panic("The " + messageType + " service is already registered")
+		return
+	}
+
+	var err error
+	service, err = ui.Register(messageType, nil, func(write func(interface{}) error) error {
+		return write(buildPeersList())
+	})
+	if err != nil {
+		log.Errorf("Unable to register Local Discovery service: %q", err)
+		return
+	}
+
+	mc = multicast.JoinMulticast()
+
+	mc.AddPeerCallback = func(peer string, peersInfo []multicast.PeerInfo) {
```

Comment:
Yeah that's why I was also suggesting adding a separate method that
internally would just call the same thing with empty functions. Again not a
big deal but to me I just think these various possible states kicking
around causes issues in the end, and in this case requires the internal
code to always check for nil on those functions.

On Fri, Jun 26, 2015 at 1:51 PM, Ulysses Aalto <notifications@github.com>
wrote:

> In src/github.com/getlantern/flashlight/localdiscovery/localdiscovery.go
> <https://github.com/getlantern/lantern/pull/2646#discussion_r33394686>:
>
> > +		panic("The " + messageType + " service is already registered")
> > +		return
> > +	}
> > +
> > +	var err error
> > +	service, err = ui.Register(messageType, nil, func(write func(interface{}) error) error {
> > +		return write(buildPeersList())
> > +	})
> > +	if err != nil {
> > +		log.Errorf("Unable to register Local Discovery service: %q", err)
> > +		return
> > +	}
> > +
> > +	mc = multicast.JoinMulticast()
> > +
> > +	mc.AddPeerCallback = func(peer string, peersInfo []multicast.PeerInfo) {
>
> The reason for this design is so the API is simpler if you don't need to
> set any of the callbacks, given that Go has no variable or named arguments.
> If you require arguments you need to provide them always, and also you have
> to check for documentation to recall the position, while if they are
> assigned to exported members of the struct you see clearly what they mean
> and if you don't add callbacks then there is no clutter.
> On the other hand, your design forces the user to really think about those
> callbacks since is forced to provide them.
> I'll change the design then.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/getlantern/lantern/pull/2646/files#r33394686>.
>



-- 
--
President
Brave New Software Project, Inc.
https://www.getlantern.org
pgp A998 2B6E EF1C 373E 723F A813 045D A255 901A FD89